_posts/2019-07-11-JS-testing-101.md

```diff
@@ -0,0 +1,433 @@
+        .finally(
+          () => {
+            expect(componentUnderTest.cars).toEqual([aRedPorsche]);
+            expect(componentUnderTest.loading).toBeFalsy();
+            done();
+          }
+        );
+
+      expect(componentUnderTest.loading).toBeTruthy();
+
+      promiseResolve([aBlackCadillac, aRedPorsche]);
+    });
+````
+
+So what’s happening here? In the promise callback, we assign the resolve function to a variable outside of that function, so we can use it later on. Then we call our method under test, but notice that we don’t use the await keyword here. Why is this? Well, if we were to use the await keyword, it would simply wait for the promise to resolve, but it would never happen because our resolve function is never called. Instead we just want to call it, let it execute and start expecting the side effects. As we saw in the method under test, the loading field is set to true before the red cars are retrieved, since those cars are not retrieved yet, we can check that the loading field is indeed set to true. Ok, now we have checked that, let's resolve the carService call. We do this by calling our promiseResolve function. Now the cars are resolved, so we can check that the loading has now been set to false and the cars field has been correctly set to the right value.
+The finally callback of the returned promise, will execute after we call our promiseResolve function, and will check the final result of our method and subsequently call the done function to tell jasmine our test is over.
+
+In the second test we will test the error scenario, where we want to see if the loading value is still set correctly and the retrievalError is also set to true.
+
+````javascript
+it('should set the retrievalError to true and loading states correctly if the promise is rejected', async done => {
+      let promiseReject: (cars: Array<Car>) => void;
+      when(carService.getCarsAsPromise()).thenReturn(new Promise((resolve, reject) => {
+        promiseReject = reject.bind(this);
+      }));
+
+      componentUnderTest
+        .getAllRedCarsAsPromiseWithStuffHappeningInBetween()
+        .finally(
+          () => {
+            expect(componentUnderTest.cars).toEqual([]);
+            expect(componentUnderTest.loading).toBeFalsy();
+            expect(componentUnderTest.carRetrievalError).toBeTruthy();
+            done();
+          }
+        );
+
+      expect(componentUnderTest.loading).toBeTruthy();
+
+      promiseReject(undefined);
+    });
+````
+
+Here we actually do the same thing, except instead of using the resolve function, we use the reject function. 
+
+### Observables
+
+For observables, I have created methods that do exactly the same as our previous promises scenarios, except we use observables here.
+
+#### Scenario A
+
+ Here we will test a simple observable that returns one value and then completes. This is a fairly common use case, especially with http calls.
```

Comment:
```suggestion
For Scenario A we will test a simple observable that returns one value and then completes. This is a fairly common use case, especially with http calls.
```
To avoid the double Here